### DISCUSSION TITLE
Tree Structural Complexity Metric

### DISCUSSION THREAD

**Discussion by Francoisdt14** [r-lidar/lidR#428] [Ideas]
Hi Jean-Romain,

I was wondering if you had any interest in adding another structural complexity metric (besides entropy/VCI)? The metric in question is called 'box dimension', and has recently been applied when investigating structural complexity in Scots pine (Saarinen et al., 2021; https://doi.org/10.1002/ece3.7216). This metric would help to understand the relationship between structural complexity and stem/crown size and shape.

The metric was first been applied to laser scanning data by Seidel in 2017 (https://doi.org/10.1002/ece3.3661), but it seems that he wrote the routine in Mathematica. The formula outlining how box dimension is calculated is outlined in section 2.1 of Seidel (2017), and section 3.2 of Saarinen et al., (2021), while the additional information and data for calculating the box dimension in Saarinen et al. can be found here:  http://doi.org/10.5281/zenodo.4419878.

I think a similar metric might be implemented using the function 'est.boxcount' in the R package 'Rdimtools' (https://github.com/kyoustat/Rdimtools, https://cran.r-project.org/web/packages/Rdimtools/Rdimtools.pdf), however they seem to use a radius rather than a box. 

Thank you for your time,

Francois

Integrated Remote Sensing Studio,
Faculty of Forestry,
UBC

**Comment by Jean-Romain** [comment]
I quickly looked into the paper and what you are talking about is the fractal dimension computed with a box counting method. I already worked on it at the very beginning of lidR but I gave up. It is relatively tricky to implement efficiently and I don't want to rely on third party package. There is a package for that https://cran.r-project.org/web/packages/fractaldim/fractaldim.pdf that provides many ways to compute the fractal dimension including boxcounting but it supports only 2D I think. `est.boxcount` seems better. Lets try it:

```r
library(Rdimtools)
cube = cbind(runif(10000), runif(10000), runif(10000))
line = cbind(0,0, runif(10000))
plane = cbind(runif(10000), runif(10000), 0)
est.boxcount(cube)$estdim
#> 2.53
est.boxcount(plane)$estdim
#> 1.81
est.boxcount(line)$estdim
#> 0.96
```

Expected values being 3,2,1 respectively, I'm not sure what is computed. But the doc makes reference to a paper. 

Anyway, once you have a function you like you can use it:

```r
library(lidR)

fd = function(X,Y,Z)
{
  M = cbind(X,Y,Z)
  est.boxcount(M)$estdim
}


LASfile <- system.file("extdata", "Megaplot.laz", package="lidR")
las = readLAS(LASfile)
metrics = grid_metrics(las, ~fd(X,Y,Z), 10)
plot(metrics, col = height.colors(50))
```

![Rplot](https://user-images.githubusercontent.com/3872279/117748485-19e18c00-b1de-11eb-8820-0c2d218267c4.jpeg)

**Reply by Francoisdt14** [reply]
Thanks for the reply! Appreciate it.

Francois

**Reply by ghost** [reply]
Thanks Jean-Romain,

I was looking for a way to implement box dimensions. Unfortunatly you are right, that this is not really efficient and takes a long time to produce a result.
However, I found the `box_counting` function in the VoxR package https://github.com/Blecigne/VoxR and seems to me much faster and produces about the same results.

```r
library(lidR)
library(VoxR)

fd = function (X,Y,Z,min_vox_size)
{
 M = data.frame(X,Y,Z)
 box_counting(M, min_vox_size = min_vox_size, message=FALSE)
}

LASfile <- system.file("extdata", "Megaplot.laz", package="lidR")
las = readLAS(LASfile)
metrics = grid_metrics(las, ~fd(X,Y,Z,1), 10)
plot(metrics, col = height.colors(50))
```

**Reply by ghost** [reply]
UPDATE:
the VoxR package seems to be abandoned and `install_github()` failed.
Therefore, the Rdimtools package recommended by JR should be used instead. At the time of writing, it is still actively being maintained.

**Comment by jpdenux** [comment]
Thank you for your developements.
They work fine with my data.

But I'm not a Lidar specialist, how this metrics can be interpreted. I'm not able to relate this calculation with the papers cited.
Any idea on the thematic meaning of this metric?

Thank you

**Reply by ghost** [reply]
According to Seidel et al. (2019) 'box-dimensions' describes the architecture of individual trees. In this code example, box-dimension is applied to grid cells and not individual trees, therefore you need to be careful with the interpretation.

Seidel et al. (2019) found that box-dimensions is related "to structural efficiency (benefit-to-cost ratio), productivity, and growing conditions (competition or availability of light)."

I hope this answers your questions.

https://onlinelibrary-wiley-com.ezproxy.utas.edu.au/doi/10.1002/ece3.5281

**Comment by NicoBBA** [comment]
I came to this thread as it seems to be closely related to what I am looking for. 

Using an area-based-approach I have been estimating forest variables such as stem volume (m3/ha) from linking ALS to sample plot measurements. However there is a discrepancy between the stem volume and the total volume of vegetation. I was hoping to find ways to deal with this using voxels, then comparing some metrics from this to the estimated stem volume and perhaps adding variability 'other vegetation volume' that I can then add in my modelling. 

Does anyone have any ideas or tips that I could look into?
